app/com/arpnetworking/metrics/portal/query/impl/KairosDbQueryExecutor.java

```diff
@@ -0,0 +1,182 @@
+import com.arpnetworking.kairos.service.KairosDbService;
+import com.arpnetworking.metrics.portal.query.QueryExecutor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import models.internal.BoundedMetricsQuery;
+import models.internal.MetricsQuery;
+import models.internal.MetricsQueryFormat;
+import models.internal.MetricsQueryResult;
+import models.internal.TimeSeriesResult;
+import models.internal.impl.DefaultMetricsQueryResult;
+import models.internal.impl.DefaultTimeSeriesResult;
+
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+
+/**
+ * A {@code QueryExecutor} that accepts KairosDB JSON metrics queries.
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public class KairosDbQueryExecutor implements QueryExecutor {
+    private final KairosDbService _service;
+    private final ObjectMapper _objectMapper;
+
+    /**
+     * Default Constructor.
+     *
+     * @param service The KairosDBService used to execute queries.
+     * @param objectMapper The objectMapper used to parse queries.
+     */
+    @Inject
+    public KairosDbQueryExecutor(final KairosDbService service, final ObjectMapper objectMapper) {
+        _service = service;
+        _objectMapper = objectMapper;
+    }
+
+    @Override
+    public CompletionStage<MetricsQueryResult> executeQuery(final BoundedMetricsQuery query) {
+        return CompletableFuture.completedFuture(query).thenCompose(this::executeQueryInner);
+    }
+
+    @Override
+    public Optional<ChronoUnit> periodHint(final MetricsQuery query) {
+        assertFormatIsSupported(query.getQueryFormat());
+        final com.arpnetworking.kairos.client.models.MetricsQuery metricsQuery;
+        try {
+            metricsQuery = _objectMapper.readValue(query.getQuery(),
+                    com.arpnetworking.kairos.client.models.MetricsQuery.class);
+        } catch (final JsonProcessingException e) {
+            throw new RuntimeException("Could not parse query", e);
+        }
+        // The period hint of the query is the smallest of each metric within
+        return metricsQuery.getMetrics()
+                .stream()
+                .map(this::periodHint)
+                .flatMap(Streams::stream)
+                .min(ChronoUnit::compareTo);
+    }
+
+    private Optional<ChronoUnit> periodHint(final Metric metric) {
+        final List<Aggregator> aggregators = metric.getAggregators();
+        if (aggregators.isEmpty()) {
+            return Optional.empty();
+        }
+        // For the purposes of periodic querying, the only aggregator size that matters
+        // is the final aggregator with a sampling parameter.
```

Comment:
Huh, I'd have guessed it would be the _first_ sampling aggregator. Or... maybe even that coarsest of _any_ sampling aggregator?

Reply:
I've been playing around with this a bit more and I think you're correct. I'll update.